### PR TITLE
Re-organized and arranged the render and _serialize methods in order to fix a bug that occured when _serialize is set to true.

### PR DESCRIPTION
**Pull request by joshrhykerd**
When _serialize was set to true in order to automatically serialize all viewVars the resulting ajax response was always _{"error":null,"content":"","1":null}_ where _serialize was set to 1 with a null value. I re-organized the code to what made a little more sense to myself as well as resolved the error.

### CODE REVIEW

**Comment by dereuromark** [comment]
Can we maybe have a test case that supports this change as a regression protection?

File: src/View/AjaxView.php [hunk 1]
```diff
@@ -95,39 +103,70 @@ public function render($view = null, $layout = null) {
...
+            foreach ($serialize as $alias => $key) {
+                if (is_numeric($alias)) {
+                    $alias = $key;
+                }
+                if (array_key_exists($key, $this->viewVars)) {
+                    $additionalData[$alias] = $this->viewVars[$key];
+                }
+            }
+
+            return !empty($additionalData) ? $additionalData : null;
```
[the remark this answers is not in the document]
Comment (joshrhykerd) [reply src/View/AjaxView.php]: Sorry I didn't pay tooooo much attention to this method as it's essentially copied from JsonView with only very minor changes.

[the remark this answers is not in the document]
Comment (dereuromark) [reply src/View/AjaxView.php]: I see, ok :) I can take a look at the end to finalize this.

**Comment by joshrhykerd** [comment]
I can look at writing a test case, but it's actually something I've never done before...

**Comment by dereuromark** [comment]
I added a bit of docs for it @ https://github.com/dereuromark/cakephp-ajax/blob/master/docs/Contributing.md
Make sure you merge in latest master, though.

File: src/View/AjaxView.php [hunk 2]
```diff
@@ -95,39 +103,70 @@ public function render($view = null, $layout = null) {
...
-	 * @param array $serialize The viewVars that need to be serialized.
-	 * @return array The serialized data.
+	 * @param array|string|bool $serialize The name(s) of the view variable(s) that
+	 *   need(s) to be serialized. If true all available view variables will be used.
+	 * @param array $additionalData Data items that were defined internally in our own
+	 *   render method.
+	 * @return mixed The data to serialize.
 	 */
-	protected function _serialize($response, $serialize) {
+	protected function _dataToSerialize($serialize = true, $additionalData = []) {
```
Comment (dereuromark) [inline src/View/AjaxView.php]: You dont need default value for $serialize, as you always pass a value in the calling part.

**Comment by dereuromark** [comment]
This starts to look really good - thx a lot. And it seems you also got a great test set up for it now. :+1:

File: src/View/AjaxView.php [hunk 3]
```diff
@@ -95,39 +103,66 @@ public function render($view = null, $layout = null) {
...
-				}
+	protected function _dataToSerialize($serialize, $additionalData = []) {
+		if ($serialize === true) {
+			$data = array_diff_key(
+				$this->viewVars,
+				array_flip($this->_specialVars)
+			);
+
+			if (empty($data)) {
+				return null;
```
Comment (dereuromark) [inline src/View/AjaxView.php]: Why is this needed? Shouldnt empty data be OK?

Comment (joshrhykerd) [reply src/View/AjaxView.php]: Yes I believe you are probably right. I will test it and make the appropriate changes. Again the majority of this method came from the JsonView so I just assumed they knew what they were doing lol.

Comment (dereuromark) [reply src/View/AjaxView.php]: OK
otherwise you need to adjust the doc block for

    protected function _serialize($dataToSerialize = []) {

to also allow null :)

Comment (joshrhykerd) [reply src/View/AjaxView.php]: The chances this conditional is met are very slim, but I guess the answer to whether it should be removed or kept comes down to what the default json_encoded ajax value should if there are indeed no viewVars. In its current state, returning null, the default ajax value is null. If we remove the conditional the default ajax value is an empty array. Which do you prefer? I guess the more I think about it the condition can never really be true because we push data into viewVars in the render method...

Comment (dereuromark) [reply src/View/AjaxView.php]: Exactly, so that case just cant happen, and array always is fine IMO.

Comment (joshrhykerd) [reply src/View/AjaxView.php]: Cool I'll get rid of it right now.

File: src/View/AjaxView.php [hunk 4]
```diff
@@ -95,39 +103,61 @@ public function render($view = null, $layout = null) {
...
+			}
+			if (array_key_exists($key, $this->viewVars)) {
+				$additionalData[$alias] = $this->viewVars[$key];
 			}
-		} else {
-			$response[$serialize] = isset($this->viewVars[$serialize]) ? $this->viewVars[$serialize] : null;
 		}
-		return $response;
+
+		return $additionalData ?: null;
```
Comment (dereuromark) [inline src/View/AjaxView.php]: That null part we also dont need then, right?

Comment (joshrhykerd) [reply src/View/AjaxView.php]: Yup your right. Not sure why I didn't catch that lol

**Comment by joshrhykerd** [comment]
What do you think about adding a another default serialized value called _status_? Client side its very common, at-least for me, to want/need to know if the ajax call was successful. Not if it completed or not, but if the action performed by the ajax call completed successfully. For example submitting a form. Did it fail validation and now an error needs shown or was it successful and redirection needs performed. I know there are other keys that you can kind glimpse that info from, but I think it would be nice to know and have a definitive value that I can always check and rely on being there that holds that information. By default it could be false OR error, and then by means of a new component method the value could be changed. Either true/false OR success/error.  I dunno just an idea...

**Comment by dereuromark** [comment]
I see that you also had to adjust the component.
Will merge this now as it is, thank you.

**Comment by dereuromark** [comment]
As for your status:
In some APIs they do that, indeed.
What do you have in mind for our situation here?
A status code (HTTP)? 200 OK, 422 validation error, ...?

**Comment by joshrhykerd** [comment]
I was kinda thinking about setting a _status_ flag of some sort in the response, but your response above got me thinking and the more I think about it HTTP status codes make more sense and really are probably APP specific and have no place here I guess...
